File: frontend/src/types/audit.ts
export type AuditActorType =
  | 'USER'
  | 'AGENT'
  | 'SYSTEM'
  | 'RUNTIME'
  | 'UNKNOWN'

export type AuditOutcome =
  | 'SUCCESS'
  | 'FAILED'
  | 'DENIED'
  | 'CANCELLED'
  | 'UNKNOWN'

import type { EntityReference, RelatedEntities } from './common';

export interface AuditActor {
  type: AuditActorType
  id?: string
  label?: string
}

export type AuditTarget = EntityReference

export interface AuditChange {
  field: string
  before?: unknown
  after?: unknown
  redacted?: boolean
}

export type AuditRelatedEntities = RelatedEntities

export interface AuditRecord {
  id: string
  timestamp: string

  actor: AuditActor
  action: string
  target?: AuditTarget

  outcome: AuditOutcome
  reason?: string

  correlationId?: string

  changes?: AuditChange[]
  related?: AuditRelatedEntities
}

export interface AuditQuery {
  search?: string
  actorType?: AuditActorType | 'ALL'
  outcome?: AuditOutcome | 'ALL'
  timeRange?: '1h' | '24h' | '7d' | 'ALL'
  correlationId?: string
}

File: backend/app/schemas/audit.py
from typing import Any, Literal, Optional
from pydantic import BaseModel, ConfigDict
from .common import EntityReference, RelatedEntities

AuditOutcome = Literal["SUCCESS", "DENIED", "FAILED"]


class AuditChangesDto(BaseModel):
    model_config = ConfigDict(extra="ignore")

    field: str
    before: Optional[Any] = None
    after: Optional[Any] = None


class AuditRecordDto(BaseModel):
    model_config = ConfigDict(extra="ignore")

    id: str
    timestamp: str
    actor: EntityReference
    action: str
    target: EntityReference
    outcome: AuditOutcome
    reason: Optional[str] = None
    changes: Optional[list[AuditChangesDto]] = None
    correlation_id: Optional[str] = None
    related: Optional[RelatedEntities] = None
